# Pandas

In [2]:
import pandas as pd


## Задачи для совместного разбора

1. Загрузите данные из файла `sp500hst.txt` и обозначьте столбцы в соответствии с содержимым: `"date", "ticker", "open", "high", "low", "close", "volume"`.

In [3]:

data = pd.read_csv('data/sp500hst.txt', sep=",", header=None)
data.columns = ["date", "ticker", "open", "high", "low", "close", "volume"]
data

date ticker   open     high     low  close  volume
0       20090821      A  25.60  25.6100  25.220  25.55   34758
1       20090824      A  25.64  25.7400  25.330  25.50   22247
2       20090825      A  25.50  25.7000  25.225  25.34   30891
3       20090826      A  25.32  25.6425  25.145  25.48   33334
4       20090827      A  25.50  25.5700  25.230  25.54   70176
...          ...    ...    ...      ...     ...    ...     ...
122569  20100813    ZMH  51.72  51.9000  51.380  51.44   14561
122570  20100816    ZMH  51.13  51.4700  50.600  51.00   13489
122571  20100817    ZMH  51.14  51.6000  50.890  51.21   20498
122572  20100819    ZMH  51.63  51.6300  50.170  50.22   18259
122573  20100820    ZMH  50.03  50.5500  49.480  49.82   17792

[122574 rows x 7 columns]

2. Рассчитайте среднее значение показателей для каждого из столбцов c номерами 3-6.

In [9]:
data.mean()[2:6]

C:\Users\tomas\AppData\Local\Temp\ipykernel_13172\131544445.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.mean()[2:6]


high         43.102243
low          42.054464
close        42.601865
volume    81395.068138
dtype: float64

3. Добавьте столбец, содержащий только число месяца, к которому относится дата.

In [19]:
data['month'] = pd.to_datetime(data['date'], format='%Y%m%d').dt.month
data

date ticker   open     high     low  close  volume  month
0       20090821      A  25.60  25.6100  25.220  25.55   34758      8
1       20090824      A  25.64  25.7400  25.330  25.50   22247      8
2       20090825      A  25.50  25.7000  25.225  25.34   30891      8
3       20090826      A  25.32  25.6425  25.145  25.48   33334      8
4       20090827      A  25.50  25.5700  25.230  25.54   70176      8
...          ...    ...    ...      ...     ...    ...     ...    ...
122569  20100813    ZMH  51.72  51.9000  51.380  51.44   14561      8
122570  20100816    ZMH  51.13  51.4700  50.600  51.00   13489      8
122571  20100817    ZMH  51.14  51.6000  50.890  51.21   20498      8
122572  20100819    ZMH  51.63  51.6300  50.170  50.22   18259      8
122573  20100820    ZMH  50.03  50.5500  49.480  49.82   17792      8

[122574 rows x 8 columns]

4. Рассчитайте суммарный объем торгов для для одинаковых значений тикеров.

In [20]:
data.groupby(['ticker'])['volume'].count()

ticker
A       245
AA      245
AAPL    241
ABC     245
ABT     245
       ... 
XTO     207
YHOO    243
YUM     245
ZION    244
ZMH     245
Name: volume, Length: 524, dtype: int64

5. Загрузите данные из файла sp500hst.txt и обозначьте столбцы в соответствии с содержимым: "date", "ticker", "open", "high", "low", "close", "volume". Добавьте столбец с расшифровкой названия тикера, используя данные из файла `sp_data2.csv` . В случае нехватки данных об именах тикеров корректно обработать их.

In [40]:
data2 = pd.read_csv('data/sp_data2.csv', sep=";", header=None)
data2.columns = [ "ticker", "company", "percentage"]
new_df = pd.merge(data, data2, on='ticker')
new_df.ticker.isnull().values.any()

False

## Лабораторная работа №2

### Базовые операции с `DataFrame`

1.1 В файлах `recipes_sample.csv` и `reviews_sample.csv` находится информация об рецептах блюд и отзывах на эти рецепты соответственно. Загрузите данные из файлов в виде `pd.DataFrame` с названиями `recipes` и `reviews`. Обратите внимание на корректное считывание столбца с индексами в таблице `reviews` (безымянный столбец).

In [43]:
recipes = pd.read_csv('data/recipes_sample.csv')
reviews = pd.read_csv('data/reviews_sample.csv')
reviews.rename( columns={'Unnamed: 0':'id'}, inplace=True )
reviews

id     user_id  recipe_id        date  rating  \
0        370476       21752      57993  2003-05-01       5   
1        624300      431813     142201  2007-09-16       5   
2        187037      400708     252013  2008-01-10       4   
3        706134  2001852463     404716  2017-12-11       5   
4        312179       95810     129396  2008-03-14       5   
...         ...         ...        ...         ...     ...   
126691  1013457     1270706     335534  2009-05-17       4   
126692   158736     2282344       8701  2012-06-03       0   
126693  1059834      689540     222001  2008-04-08       5   
126694   453285  2000242659     354979  2015-06-02       5   
126695   691207      463435     415599  2010-09-30       5   

                                                   review  
0       Last week whole sides of frozen salmon fillet ...  
1       So simple and so tasty!  I used a yellow capsi...  
2       Very nice breakfast HH, easy to make and yummy...  
3       These are a favorite for the holidays and so e...  
4       Excellent soup!  The tomato flavor is just gre...  
...                                                   ...  
126691  This recipe was great! I made it last night. I...  
126692  This recipe is outstanding. I followed the rec...  
126693  Well, we were not a crowd but it was a fabulou...  
126694  I have been a steak eater and dedicated BBQ gr...  
126695  Wonderful and simple to prepare seasoning blen...  

[126696 rows x 6 columns]

1.2 Для каждой из таблиц выведите основные параметры:
* количество точек данных (строк);
* количество столбцов;
* тип данных каждого столбца.

In [45]:
recipes.info()
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   name            30000 non-null  object 
 1   id              30000 non-null  int64  
 2   minutes         30000 non-null  int64  
 3   contributor_id  30000 non-null  int64  
 4   submitted       30000 non-null  object 
 5   n_steps         18810 non-null  float64
 6   description     29377 non-null  object 
 7   n_ingredients   21120 non-null  float64
dtypes: float64(2), int64(3), object(3)
memory usage: 1.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126696 entries, 0 to 126695
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   id         126696 non-null  int64 
 1   user_id    126696 non-null  int64 
 2   recipe_id  126696 non-null  int64 
 3   date       126696 non-null  object
 4   rating     126696 non-nul

1.3 Исследуйте, в каких столбцах таблиц содержатся пропуски. Посчитайте долю строк, содержащих пропуски, в отношении к общему количеству строк.

In [49]:
print(recipes.isnull().mean() * 100)
reviews.isnull().mean() * 100

name               0.000000
id                 0.000000
minutes            0.000000
contributor_id     0.000000
submitted          0.000000
n_steps           37.300000
description        2.076667
n_ingredients     29.600000
dtype: float64


id           0.000000
user_id      0.000000
recipe_id    0.000000
date         0.000000
rating       0.000000
review       0.013418
dtype: float64

1.4 Рассчитайте среднее значение для каждого из числовых столбцов (где это имеет смысл).

1.5 Создайте серию из 10 случайных названий рецептов.

1.6 Измените индекс в таблице `reviews`, пронумеровав строки, начиная с нуля.

1.7 Выведите информацию о рецептах, время выполнения которых не больше 20 минут и кол-во ингредиентов в которых не больше 5.

### Работа с датами в `pandas`

2.1 Преобразуйте столбец `submitted` из таблицы `recipes` в формат времени. Модифицируйте решение задачи 1.1 так, чтобы считать столбец сразу в нужном формате.

2.2 Выведите информацию о рецептах, добавленных в датасет не позже 2010 года.

### Работа со строковыми данными в `pandas`

3.1  Добавьте в таблицу `recipes` столбец `description_length`, в котором хранится длина описания рецепта из столбца `description`.

3.2 Измените название каждого рецепта в таблице `recipes` таким образом, чтобы каждое слово в названии начиналось с прописной буквы.

3.3 Добавьте в таблицу `recipes` столбец `name_word_count`, в котором хранится количество слов из названии рецепта (считайте, что слова в названии разделяются только пробелами). Обратите внимание, что между словами может располагаться несколько пробелов подряд.

### Группировки таблиц `pd.DataFrame`

4.1 Посчитайте количество рецептов, представленных каждым из участников (`contributor_id`). Какой участник добавил максимальное кол-во рецептов?

4.2 Посчитайте средний рейтинг к каждому из рецептов. Для скольких рецептов отсутствуют отзывы? Обратите внимание, что отзыв с нулевым рейтингом или не заполненным текстовым описанием не считается отсутствующим.

4.3 Посчитайте количество рецептов с разбивкой по годам создания.

### Объединение таблиц `pd.DataFrame`

5.1 При помощи объединения таблиц, создайте `DataFrame`, состоящий из четырех столбцов: `id`, `name`, `user_id`, `rating`. Рецепты, на которые не оставлен ни один отзыв, должны отсутствовать в полученной таблице. Подтвердите правильность работы вашего кода, выбрав рецепт, не имеющий отзывов, и попытавшись найти строку, соответствующую этому рецепту, в полученном `DataFrame`.

5.2 При помощи объединения таблиц и группировок, создайте `DataFrame`, состоящий из трех столбцов: `recipe_id`, `name`, `review_count`, где столбец `review_count` содержит кол-во отзывов, оставленных на рецепт `recipe_id`. У рецептов, на которые не оставлен ни один отзыв, в столбце `review_count` должен быть указан 0. Подтвердите правильность работы вашего кода, выбрав рецепт, не имеющий отзывов, и найдя строку, соответствующую этому рецепту, в полученном `DataFrame`.

5.3. Выясните, рецепты, добавленные в каком году, имеют наименьший средний рейтинг?

### Сохранение таблиц `pd.DataFrame`

6.1 Отсортируйте таблицу в порядке убывания величины столбца `name_word_count` и сохраните результаты выполнения заданий 3.1-3.3 в csv файл. 

6.2 Воспользовавшись `pd.ExcelWriter`, cохраните результаты 5.1 и 5.2 в файл: на лист с названием `Рецепты с оценками` сохраните результаты выполнения 5.1; на лист с названием `Количество отзывов по рецептам` сохраните результаты выполнения 5.2.